### DISCUSSION TITLE
How can I make multiple HTTP requests in parallel?

### DISCUSSION THREAD

**Discussion by itsbe-au** [zauberzeug/nicegui#969] [Q&A] [answered]
### Question

First of all - well done on making NiceGUI, it's incredible, and I hope I'll be in a position to contribute to its development very very soon!

In fact, I'm so impressed that I'm already challenging myself to port over an internal Next.js app to NiceGUI, and so far, it's going really nicely. One thing I am stuck on though is how to make multiple concurrent API requests.

For example: I need to make a few requests that return:

* A list of countries
* A list of organisations
* A list of standards and standard versions.

This is easy enough, if a little cumbersome, in my current app:

```javascript
const {data: standardData} = useSWR(`${process.env.API_HOST}/standard`, fetcher);
const {data: standardVersionData} = useSWR(`${process.env.API_HOST}/standard-version`, fetcher);
const {data: countryData} = useSWR(`${process.env.API_HOST}/country`, fetcher);
const {data: memberData} = useSWR(`${process.env.API_HOST}/member`, fetcher)
```

(or a similar thing with `useState` and `useEffect` to run these queries when the component is loaded).

The closest thing I can think to do in NiceGUI is this:

```python
class FormOptions:
    def __init__(self):
        self.countries = None
        self.members = None
        self.standards = None
        self.standard_versions = None

form_options = FormOptions()

async def get_countries():
    data = requests.get(f'{API_ROOT}/countries')
    form_options.countries = data.json()
    print(form_options.countries)

async def get_standards():
    data = requests.get(f'{API_ROOT}/standards')
    form_options.standards = data.json()
    print(form_options.standards)

app.on_connect(get_countries())
app.on_connect(get_standards())
# ...etc
```

But, the problem is I don't want to load all this data when the **app** itself is loaded – only when the relevant page is accessed.

Any ideas?

**Comment by rodja** [comment, marked as answer]
Thanks for the praise. We would love to see you contributing. About your question:

NiceGUI (and the underlying FastAPI) are async frameworks. That means, no io-bound or cpu-bound tasks should be directly executed on the main thread but rather be "awaited". See https://fastapi.tiangolo.com/async/ for more details. The popular `requests` lib unfortunately has no async support. I recommend you switch to [httpx](https://www.python-httpx.org/async/). With [`asyncio.gather`](https://docs.python.org/3/library/asyncio-task.html#asyncio.gather) you can await multiple coroutines simultaneously.

**Reply by itsbe-au** [reply]
Thanks @rodja for the super quick reply - yes of course, very valid point to use `httpx` instead.

So I guess I actually don't need anything like `app.on_connect(...)`, in that case? Just gathering and awaiting those coroutines under a `@ui.page(...)` should be enough, right?

**Reply by rodja** [reply]
Exactly. Yesterday I wrote something very similar: https://github.com/zauberzeug/nicegui/blob/9ad5f4180cecde693484946d3484c735a73571e6/examples/lightbox/main.py#L42-L46
